File: code/01_workshop.py
# 01_workshop.py
# Tim Fraser
# Workshop 1: Coding in Python
# Chapter: Coding in Python


# Getting Started ---------------------------------------------------------

## Installing packages ----
# Upgrade pip
# !python -m pip install --upgrade pip
#
# This course pins its Python packages in functions/requirements.txt.
# Install them all at once:
# !pip install -r functions/requirements.txt
#
# Or install the main packages for this course one at a time:
# !pip install pandas==2.3.3
# !pip install scipy==1.16.3
# !pip install statsmodels==0.14.5
# !pip install patsy==1.0.2
# !pip install plotnine==0.15.1
# !pip install dfply==0.3.3

# Print hello world!
print("hello world")

## Importing packages ----
import pandas as p # Import pandas
from dfply import * # Import all dfply functions
from plotnine import * # Import all plotnine functions


# Basic Calculations ------------------------------------------------------

# Addition
print(1+2)

# Vector
print([1,2,3,4,5])

# Subtraction
print(5 - 2)

# Multiplication
print(2 * 3)

# Division
print(15 / 5)

# Exponents
print(2**2)

# Square Root
print(16**0.5)

# Order of Operations
print(2 * 2 - 5)

# Use parentheses!
print(2 * (2 - 5))


# Eg.
print(64**.5)
print(64**(1/3))
print(8**2)


# Types of Data in Python -------------------------------------------------

## Values and variables ----

print(2) # this is a value
print("x") # this is a value

myvalue = 2

secondvalue = myvalue + 2

# In Positron, you can just print the values to console,
# without using the print() command.
print(secondvalue)


## Lists (like R vectors) ----
print([1,2,3])
print(["Boston", "New York", "Los Angeles"])


# Here's a vector of (hypothetical) seawall heights in 10 towns.
myheights = [4, 4.5, 5, 5, 5, 5.5, 5.5, 6, 6.5, 6.5]

# And here's a list of hypothetical names for those towns
mytowns = ["Gloucester", "Newburyport", "Provincetown",
             "Plymouth", "Marblehead", "Chatham", "Salem",
             "Ipswich", "Falmouth", "Boston"]

# And here's a list of years when those seawalls were each built.
myyears = [1990, 1980, 1970, 1930, 1975, 1975, 1980, 1920, 1995, 2000]

# To manipulate them, we'll need to bundle them into pandas objects.


## DataFrames with pandas ----

# let's bundle them into a data.frame with pandas.
sw = p.DataFrame({'height': myheights, 'town': mytowns, 'year':myyears})
# Add 2 to all the heights
print(sw.height + 2)

# Or just make it a series, and then add 2.
print(p.Series(myheights) + 2)


# Element-wise multiplication
print(p.Series(myheights) * p.Series(myheights))
 # or
print(sw.height * sw.height)

# Matrix Multiplication...
print(sw.height.dot(sw.height))


# Common Functions in Python ----------------------------------------------

# Descriptive Stats
print(sw.height.mean())
print(sw.height.median())
print(sw.height.min())
print(sw.height.max())
print(sw.height.mode())
print(sw.height.quantile(q = 0.5))


# The Pipeline ------------------------------------------------------------

# Here's a brief test of using dplyr-style functions
# with dfply
# diamonds is a dataset loaded within dfply

## Select, mutate, summarize ----

# Select just one column
print(sw >> \
  select(X.height))

# Mutate a column
print(sw >> \
  mutate(y = X.height ** X.height))


# Summarize a data.frame
print(sw >>\
  summarize(mean_value = mean(X.height) ))

## Grouping, arranging, filtering ----

# Get the mean price per diamond cut
print(diamonds >>\
  group_by(X.cut) >>\
  summarize(price = mean(X.price )))

print(diamonds >>\
  arrange(X.price, ascending=False) >>\
  head())


# Filtering
print(diamonds >>\
  mask(X.carat < 0.23) >>\
  head())


print(diamonds >>\
  rename(CUT=X.cut, COLOR='color'))


print(diamonds >>\
  arrange('color'))

print(diamonds >>\
  group_by('cut') >>\
  summarize(price = mean(X.price)))

## Reshaping ----

# gather() stacks several columns into key-value pairs.
# Unlike R's tidyr, dfply's gather() requires you to name
# the key column, the value column, and the columns to stack.
print(diamonds >>\
  gather('measure', 'value', ['x', 'y', 'z']) >>\
  head())


# Clear environment
globals().clear()

File: code/01_workshop.R
# 01_workshop.R
# Tim Fraser
# Workshop 1: Introduction - basic operations in R
# Chapter: Coding in R
#
# Check out timothyfraser.com/sigma for more.

# Let's load some packages
# install.packages("dplyr") # install dplyr
# install.packages("ggplot2") # install ggplot2
library(dplyr) # turn on dplyr
library(ggplot2) # turn on ggplot2

# Comments! 
# This is addition!
1+2

1+2
1*5
2^3
(2*3)
((2*2)+3)

4^0.5
4^(1/2)
4^(1/3)
sqrt(4)





myobject = c(1,2,3,4,5)

myobject

myobject + 1

myobject *2

# Heads up: this next line is SUPPOSED to fail. Run it anyway, and read what R
# says back to you - you can't add 1 to words. That error message is the lesson.
c("corgi", "dalmatian", "terriers") + 1
c("corgi", "dalmatian", "terriers", 1)



d = data.frame(
  doggies = c("corgis", "dalmatians", "terriers"),
  count = c(5, 3, 2)
)

d$count
d$count + 1
d$count + d$count

d$count = d$count + 1

# Not this!
# d = d$count + 1

# dataframe[row,column]
d[1,1]
d[3,2]
d[3, ]
d[, 2]

# adding columns
d$weight = c(100, 40, 20)

d$weight2 = c(100, 40, NA)

# using dplyr function bind_rows()
bind_rows(
  data.frame(x = c(1,2), y= c(3,4)),
  data.frame(x=  c(3,4), y = c(5,6))
)
bind_rows(
  data.frame(x = c(1,2), y= c(3,4), z= c(3,4)),
  data.frame(x=  c(3,4), y = c(5,6))
)
# from dplyr package
tibble(
  x = c(1,2), y= c(3,4), z= c(3,4)
)


d$weight2 = NULL

# Heads up: this next line is SUPPOSED to fail too. We just deleted weight2
# above, so there is nothing left to deselect. R tells you exactly that:
# "Can't select columns that don't exist." Deleting a column is permanent -
# that error is how you find out you already did it.
d %>% select(-weight2)






c(1,2,3,4)

# Heads up: this next line is SUPPOSED to fail, for the same reason as the one
# up at the top. Multiplying is arithmetic, and these are words - it fails no
# matter which arithmetic you try. Compare it to the line just below it.
c("corgi", "dalmatian") * 2

c(1,2,3,4) * 2

c(1,2,3,4) * c(1,2,3,4)

c(1,2,3,4) %*% c(1,2,3,4)


sqrt(64)
64^(1/3)
8^2






coffee = c(2, 4, 5, 6, 7, 3,2, 3,4)
coffee

sales = c(3.4,2.5, 3.2, 6.3, 4, 3, 6, 7, 8)
sales

coffee * 2

# coffees per dollar
coffee / sales
# dollars per coffee
sales / coffee



# Let's try working with data.frames.

data.frame(coffee, sales)

dat = data.frame(coffee = coffee, sales = sales)

dat

# We can index specific values, rows, and columns...
dat[,1]
dat[1,]
dat[1:3,]
dat[, 1:2]




# And we can use dplyr functions to do actions to dataframes.
#    %>%    



select(dat, coffee)
dat %>% select(coffee)

head(dat, 2)
dat %>% select(coffee) %>% head(2)


dat %>% 
  select(coffee) %>% 
  head(2)


dat %>%
  select(coffee) %>%
  slice(1:2)

dat %>%
  select(coffee) %>%
  filter(coffee > 4)


dat %>%
  filter(coffee == 4)
dat %>%
  filter(coffee >= 4)
dat %>%
  filter(coffee <= 4)

dat %>%
  filter(coffee %in% c(4, 5))

dat %>%
  summarize(avg = mean(coffee))

# find the average and the standard deviation - yeah!
dat %>%
  summarize(avg = mean(coffee),
            stdev = sd(coffee))

data.frame( 
  adorable = c(1, 0),
  glasses = c("yes", "no")
)

# get 2000 values
nums = data.frame(x = 1:2000)
nums # only shows first 1000 values
# if you assign you get no output...
num2 = nums %>% slice(1000:1003)


# Clear my environment
rm(list = ls())

# Here's a brief test of using dplyr-style functions
# with dplyr 
# diamonds is a dataset loaded within ggplot2
library(dplyr)
library(ggplot2)

# Get the mean price per diamond cut
diamonds %>%
  group_by(cut) %>%
  summarize(price = mean(price ))

# Arranging
diamonds %>%
  arrange(price) %>%
  head()


# Filtering
diamonds %>%
  filter(carat < 0.23) %>%
  head()


diamonds %>%
  rename(CUT = cut, COLOR='color')


diamonds %>%
  arrange(color)

diamonds %>%
  group_by(cut) %>%
  summarize(price = mean(price))



# We can clean up using rm()
rm(list = ls())
